compiler/rustc_privacy/src/lib.rs

```diff
@@ -1188,6 +1188,15 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
         self.maybe_typeck_results = old_maybe_typeck_results;
     }
 
+    fn visit_generic_arg(&mut self, generic_arg: &'tcx hir::GenericArg<'tcx>) {
+        match generic_arg {
```

Comment:
this is what's currently broken here. We look at the final `ty::Ty` for the generic arg in `visit_ty` which can be relevant even for `GenericArg::Infer`. *I also don't know why we still walk the `hir_ty` after visiting the `ty::Ty` in there, but that's probably a more complex change if it's even possible/desirable.*

This happens after typecheck, so we can get the expected generic arg here. We probably want to put this into a separate query. Something like the following which always returns either a `GenericArg::Type` or `GenericArg::Const`:
```rust
query generic_arg_for_infer_arg(id: HirId) -> hir::GenericArg
```
I think ideally we would use this query by default and never actually implicitly ignore the ty/const in `visit_generic_arg`. This would require visitors without a `tcx` and which don't care about inference types/const to manually overwrite `visit_generic_arg`.

@varkor @petrochenkov What's your opinion here?

Reply:
I agree we never want to be ignoring the type/const. Having a query seems like a reasonable solution.

Reply:
Hm, after looking at this for a bit, it seems like a significant change to the whole structure, but this is the only visitor which seems to have this problem. The typeck results are actually available at that point, but for some reason the inference variable doesn't seem to be in them, so I was wondering if I could just modify the typeck function instead?